### DISCUSSION TITLE
Partnership with the Bootstrap team

### DISCUSSION THREAD

**Discussion by thiagomajesk** [ThinkUniform/uniformcss#11] [General]
First of all, cool project! 

After using TailwindCSS for quite some time, I always found myself at some point, wanting a little bit more control over CSS generation - given that Sass has a lot of niceties (mainly not having to rely on Javascript) it's usually far more intuitive. 

_(BTW, I also had a [pretty similar idea](https://github.com/thiagomajesk/bootilities) quite some time ago, just didn't have the time or patience to follow it through)._

So, I wanted to ask... Have you considered reaching out to [@mdo](https://github.com/mdo) in order to establish some sort of partnership for the future versions of Bootstrap? I ask that because Bootstrap has been playing around with utilities since V3 and in V5 they finally introduced their own utility API (which is still very limited in scope). 

I can almost see a future where Bootstrap (perhaps V6) will decouple its utility generation API into another project and I would see this as a complete win for both sides and an advantage over the current Tailwind "ecosystem" if we got a strong foundation based on utilities. I say this because people often choose Bootstrap and make a customization/ freedom sacrifice over the cost of building components upfront, while the Tailwind approach goes the other way around.

**Comment by thiagomajesk** [comment]
Hi @jinsupark! Could you share some thoughts on this?

**Reply by jinsupark** [reply]
Heya @thiagomajesk just catching up on this, thanks for reaching out!

First of all, thanks for checking out this project. Yes, that is precisely why I wanted to build this project as well, I found that there were certain naming conventions that were unintuitive and also saw an opportunity to provide more customizability (Sass already provided so much benefit, I wanted to build from this).

I've always thought of Uniform CSS to be a direct Sass competitor of Tailwind but I have never thought about partnering with Bootstrap. Thanks! This is a very interesting idea.

Do you have any contacts from the Bootstrap team for an intro? If not, I will reach out to @mdo to see if the Bootstrap team would be willing to discuss future opportunities where we could possibly collaborate together.

More specifically could you mention some of the limitations in scope you've faced using the Bootstrap utilities API?

**Reply by thiagomajesk** [reply]
Hi man, I'm glad to hear that we share some of those opinions, Sass is a really good tool (a little underappreciated IMHO) so I'd love to spread the word about tools that are trying to explore it and bring more options to the table.

One thing that Tailwind and Uniform both do very well is establishing solid conventions and I think that if we can leverage this by having some intersection with Bootstrap it could bring a lot of power to users.

I don't know anyone from the core team personally, but I'm trying to speak my mind and see if anyone buys the idea. You can read more about it on these threads: https://github.com/twbs/bootstrap/discussions/35014 and https://github.com/twbs/bootstrap/discussions/33885 (@mdo seems very open to discussion though).

> More specifically could you mention some of the limitations in scope you've faced using the Bootstrap utilities API?

I opened [this issue](https://github.com/twbs/bootstrap/issues/32183) some time ago and it goes into a little more detail about the problems I had with it. It's mostly about controlling how utilities are generated and how the current API is very strict to generate "Bootstrap compatible utilities".

I'd even say that the main problem is towards the frameworks naming conventions that mix component and utilities sizing. For example, if you take [this utility](https://github.com/twbs/bootstrap/blob/main/scss/_utilities.scss#L56) and try to apply the `responsive` option it will generate something like: `shadow-sm-sm`, `shadow-sm-lg` and so on (for each breakpoint and variation). 

There are other restrictions if you want to deal with utilities that have implicit CSS variables like for text opacity, gradients and other things that Tailwind and Uniform support out of the box simply because there's no way to achieve that with the API. I think that it is small in scope purposefully because Bootstrap is a "mixed framework".

IMHO this could easily be solved by generating utilities with breakpoints like Tailwind and Uniform does - by prefixing the class name with `sm:shadow-sm` or `sm.shadow-sm`. After working on my project for a while I stumbled upon a lot of cases where I had to come up with creative solutions so classes wouldn't end up with weird names: https://github.com/thiagomajesk/bootilities#notable-differences.

**Comment by jinsupark** [comment]
Hey, @thiagomajesk thank you for sharing your experience and thoughts! 

Sorry for the delayed response, we've been busy making some big improvements and finishing the documentation site. Now that we have reached this milestone, I am reaching out to you to let you know I will attempt to get in touch with @mdo to see if there are any opportunities we could work together to improve the benefits of Bootstrap and Uniform even further.

To give you a quick update on what's been happening, tutorial videos are now recorded and published, documentation is now polished and complete, and Uniform CSS is now officially launched as a fully-featured stable product on [Product Hunt](https://www.producthunt.com/posts/uniform-css)!

A big part of Uniform's philosophy is making sure customizations can be made to almost every aspect of a utility class. With that being said, Uniform CSS can be considered as both a "CSS framework" and a "utility generator". Since by nature, utility classes are immutable and designed to do one thing, we follow a very strict naming convention to ensure each class name is named consistently and concisely so that they are scalable, memorable, consistent, and future-proof.

We gathered some great feedback and launched some cool features in v1.2 that pushes the flexibility and philosophy of this framework even further:
- You can now export all your utilities and their settings as a JSON format so that you can build your class reference documentation.
- Custom utilities can now feature multiple properties.
- Important can be enabled on a per utility basis.
- You can now extract components using the `apply()` mixin similar to Tailwind.

To learn more please check out the [documentation](https://uniformcss.com/) or read the [announcement](https://github.com/ThinkUniform/uniformcss/discussions/19).

**Reply by thiagomajesk** [reply]
No worries at all man, I'm also kinda swamped until the holidays, so I've not been able to keep up with a lot.
Congratulations on the official release! It's good to know that you've been working on the project (very cool stuff BTW). I'll take a look at the new documentation as soon as I can, hope to see progress towards that front with bootstrap 🤞.

**Reply by mdo** [reply]
I've been on medical leave from work and mostly heads down on open source the last few months, but I'm slowly coming back up for air :). I'm not sure what if anything you have in mind, but I'm always down to chat! Shoot me a DM on Twitter or send me an email if you'd like :).

And a belated congrats on the launch! 🎉

**Reply by thiagomajesk** [reply]
Jeez, @mdo, I hope you're ok  (also on the health side - jk haha 😁).

**Reply by mdo** [reply]
I really don’t appreciate the joke @thiagomajesk 😕

**Reply by thiagomajesk** [reply]
Really!? Perhaps this is a cultural barrier type of thing (since I'm not a native English speaker), perhaps it didn't translate as well as I intended... I'm positive someone might think it was a little bit funny, though - just a silly bachelor-like type of joke.  Just to be sure: no offense intended my man, congratulations on your wedding, and cheers to your health pal 😉🎉.

**Reply by jinsupark** [reply]
Hey, @mdo I hope you are making a fast recovery, and thanks for checking out this project! Glad to hear you are slowly coming back up. I'd love to have the opportunity to connect with you, I've sent you a DM on Twitter 👍
